### ISSUE TITLE
Allow using baseline for font marker placement

### ISSUE THREAD

**Issue by m-kuhn** [qgis/QGIS#59732]
### What is the bug or the crash?

It should be possible to use the baseline of a font for symbol placement.

The following image shows the font [cadastra](https://www.cadastre-manual.admin.ch/de/plan-fur-das-grundbuch) which has set the baseline to the center of the symbol. The point should be centered over the intersection point of the lines.

![image](https://github.com/user-attachments/assets/b8b0a5ab-d82a-4c9f-95b7-ce919791e586)

This is currently possibly by setting an offset, but this becomes complex with zoom levels and units

![image](https://github.com/user-attachments/assets/555b8c77-988d-4903-9446-b7ba8a4824d7)

The font would be much more usable if the baseline could be set as reference for placement.

### Steps to reproduce the issue

Download the font [cadastra](https://www.cadastre-manual.admin.ch/de/plan-fur-das-grundbuch)
Configure it as symbol marker as shown above

### Versions

3.40

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

**Comment by nyalldawson** [comment]
Ideally we'd have a mode where the character bounds are calculated, and then the center of these used as the "origin" for the symbol...

I think this would need to be an option though, or we'd break a lot of existing symbology

**Comment by nirvn** [comment]
@nyalldawson , was discussing this with @m-kuhn not too long ago and we both reached the same conclusion.

**Comment by m-kuhn** [comment]
> Ideally we'd have a mode where the character bounds are calculated, and then the center of these used as the "origin" for the symbol...

This is yet another option for placement (which happens to work equally in this specific scenario on the image), no?

> I think this would need to be an option though, or we'd break a lot of existing symbology

Agreed, the question is if we can extend the combobox for vertical placement or if we need a new checkbox or similar

@signedav I think you already had some thoughts on this?

**Comment by signedav** [comment]
Yes. Since I'm not working today I'll share em tomorrow and will much appreciate your feedback then.

**Comment by signedav** [comment]
Summarizing my notes (you might know all this already)...

Currently we do have three modes as vertical anchor point:

![Screenshot from 2024-12-05 12-59-57](https://github.com/user-attachments/assets/315899d4-f596-4239-8d12-4d609488d546)

The **middlepoint** is everywhere the **source of truth**: it's calculated in the font marker layer as the middle of the _baseline_ and _QFontMetrics::ascent()_ , what is basically the highest possible point of a font (like the top of the 'aigu' of the È). 

Still when a letter is vertically centered with `VCenter` it looks mostly like this (example with Deja Vu):
![centered](https://github.com/user-attachments/assets/564e1183-e16b-4ee2-9775-17e344fff4f4)

Because it's the middle of the ascent-baseline and not the middle of the characterize, it's not really centered on some letters:

![image](https://github.com/user-attachments/assets/5fd8ca04-7250-4b59-990d-9d2a7daafbfd)
(while the À is correct, the A is not)

This would require the character bounds to fix, but that's not the main topic here anyway.

Concerning this cadastre-font to have all symbols centered according to the sign, it would not solve the issue. We just want to have them right on the baseline. 

We might assume, it should set the symbols on the right place with the setting "Bottom":
![baseline](https://github.com/user-attachments/assets/fb4b2d45-303a-4035-85fd-4d1b975d96b3)

**But it doesn't do it.**

Because we make the following calculation:
- Calculate the middlepoint like above: (ascent-baseline-difference) / 2
- Then we get the symbol size (the one from the settings, scaled depending on the unit and are dividing with 2 -> having an offset
- And we set this offset to the middlepoint.

This would work, when the size would fit the ascent-baseline, but it does not. The ascent-baseline is smaller, what means the letter is slightly above:

![image](https://github.com/user-attachments/assets/15e86860-b8ec-45da-a78e-b8da02465d25)

Instead of expected this:
![image](https://github.com/user-attachments/assets/6c7d4391-6084-47f4-8ac8-2031265f0fad)

And on the cadastre-font it looks like this:
![image](https://github.com/user-attachments/assets/69641f93-464d-4c8f-81c2-2913c718f8c8)

Instead of expected this:
![image](https://github.com/user-attachments/assets/d386bd12-1e0f-4674-91ce-3093ee087655)

**To fix this we should calculate the middlepoint according to the scaled fontsize or calculate the offset according to the ascent-baseline.**

And yes, I guess it needs a setting because otherwise it breaks existing projects.

Pro for another anchorpoint option in the combobox would be that it's more elegant (having e.g. "Baseline" there). Con is, that we kind of loose the setting when switching between the markers. This would be solved by a new checkbox "Consider Baseline as Bottomline".

Not sure how much "VCenter" or "Top" could be improved here, but I think on "Top" we could make an improvement as well that it does set the offsets in the same way like on "Bottom". If so, I would prefer a checkbox "Consider Baseline" over two new anchor point types. But concerning GUI I have no strong opinion.

**Comment by m-kuhn** [comment]
Thanks for the explanation.
For my understanding: if we go for a new checkbox "Use baseline as bottom line", activating it will change the calculated position for "Bottom" and "Top" (but `VCenter` will be unchanged)?

**Comment by signedav** [comment]
Maybe some ideas for the `VCenter` topic, even though **it does not concern this specific issue with the cadastre symbols**.

Currently it's centered according the size of the biggest character with accent etc:
![current](https://github.com/user-attachments/assets/7dccbde3-e8ab-45a4-b10a-2197564037fc)

When it would center it according to the bounds of the current symbol it would look like this:
![bounds](https://github.com/user-attachments/assets/423512d3-c3f4-4eb5-8498-5d91ed4bce26)

And something completely different would be that it's centered according to a accent-less capitol letter like this:
![caps](https://github.com/user-attachments/assets/328f6e76-2937-4baf-b510-987633092121)

**Comment by nyalldawson** [comment]
Backing up a step, from @m-kuhn:

> This is yet another option for placement (which happens to work equally in this specific scenario on the image), no?

I'd say this particular font is an unusual situation (centering characters over the baseline). But there's MANY MANY symbol fonts which don't do this which would benefit from a proper "center over the actual character bounds" placement mode. (Eg, all the ESRI symbol fonts, which have characters like this:

![image](https://github.com/user-attachments/assets/92c77e6d-7059-4083-85a2-0ea5f51b0b17)

)

Couldn't BOTH situations be fixed with a "center marker over character bounds" mode? 

My preference would be for the existing modes in the Anchor Point combo (for font markers only!) to be renamed to "Top (Legacy)", "Bottom (Legacy)" "VCenter (legacy)" and then new modes added for "Top", "Middle", "Bottom" which are calculated based on the actual character bounds. If there's still a specific requirement for a "baseline" mode then that could also be added to this same combo box.

All NEW font marker symbol layers would default to "Middle", so only existing ones would use the "VCenter (legacy)" mode...

**Comment by signedav** [comment]
Thanks for your reply. 

> I'd say this particular font is an unusual situation

Yes. But I guess the reason is that the point where the symbol should be centered to is *not* the center of the symbol bound.

![Screenshot_20241211-071001_1](https://github.com/user-attachments/assets/7bd28f29-5307-43ef-8656-6b6ab89f2546)

... and thats - as far as I understand the bound-approach - why it wont solve it for this.

Wouldn't the baseline approach make sense for other *normal* fonts as well? Like here:

![392809369-6c7d4391-6084-47f4-8ac8-2031265f0fad](https://github.com/user-attachments/assets/352e0010-a3af-4092-9af0-934bda9a2347)

**Comment by nyalldawson** [comment]
@signedav 

> Wouldn't the baseline approach make sense for other normal fonts as well? Like here:

Yes, you're correct. I'd still argue that we also need methods which consider the bounds to be useful for the majority of symbol fonts though :+1:

**Comment by signedav** [comment]
Alright, I do agree that the reference to the individual bounds would make sense. Not only for the symbol fonts.

So best would be to have an option to have the **symbol bounds** as reference and to have the **font baseline** as reference. And of course the third option of the inconsistent **current way to do it**.

![image](https://github.com/user-attachments/assets/d96a5bff-a3c0-4ff5-89ea-2aa4a694a296)

The **symbol bounds** would lead to (vertical) placement like this:
![bounds](https://github.com/user-attachments/assets/3e60c6df-baeb-4da4-b8da-230a960b4b40)

And the **baseline** would lead to (vertical) placement like that:
![baseline](https://github.com/user-attachments/assets/fa6ca0b0-7dce-4b6b-b483-c85532d24159)

And the **None (legacy)** would lead to what it does currently: Place above (top) the position of the highest bound in the font plus (difference between highest point and baseline and the size of the font divided with two*) and below (bottom) of the baseline minus (difference between highest point and baseline and the size of the font divided with two*) and the center is like with the baseline.

\* what is the current "bug". See https://github.com/qgis/QGIS/issues/59732#issuecomment-2520242221

Means both (baseline and bounds) would make sense to implement, still if only one of them will be implemented, it might make sense to have the setting prepared as a combobox (enum) instead of a checkbox (bool). 

Does this make sense for you @nyalldawson @nirvn @m-kuhn ?

P.S. having a meeting with the potential sponsor next week - will keep you up to date.

**Comment by signedav** [comment]
No strong opinion.

Your suggestion means:
- top legacy
- top bounds
- center legacy
- center bounds
- bottom legacy
- bottom bounds
- bottom baseline

Or wont you make top and bottom bounds?

**Comment by nyalldawson** [comment]
@signedav 

> Your suggestion means:

> top legacy
top bounds
center legacy
center bounds
bottom legacy
bottom bounds
bottom baseline

Exactly! (And "center bounds" would become the new default)
